### PR TITLE
Fix assistant masks for processors

### LINKED ISSUE #44521: apply_chat_template returns all-zero assistant_masks for multimodal inputs
### System Info

transformers==5.3.0

### Who can help?

@ArthurZucker and @itazap

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoProcessor

messages = [
    dict(
        role="user",
        content=[
            dict(type="image", image="test.jpg"),
            dict(type="text", text="Describe the image above."),
        ],
    ),
    dict(
        role="assistant",
        content=[
            dict(type="text", text="The image above shows a cat sitting on a table."),
        ],
    ),
]

processor = AutoProcessor.from_pretrained("Qwen/Qwen3.5-0.8B")
inputs = processor.apply_chat_template(
    messages,
    tokenize=True,
    return_dict=True,
    return_tensors="pt",
    return_assistant_tokens_mask=True,
)

print(inputs["assistant_masks"])
```

### Actual behavior

assistant_masks is all zeros.

### Expected behavior

The tokens corresponding to the assistant response should be marked with 1 in assistant_masks.

### Suspected cause
I think the issue comes from a mismatch between generation_indices and the final tokenized prompt in multimodal cases.

In AutoProcessor.apply_chat_template, the assistant mask is computed roughly like this:
```python
prompt, generation_indices = render_jinja_template(
    conversations=conversations,
    chat_template=chat_template,
    **template_kwargs,
    **special_tokens_map,
)

...

out = self(
    text=prompt,
    images=batch_images if images_exist else None,
    videos=batch_videos if videos_exist else None,
    audio=batch_audios if batch_audios else None,
    **kwargs,
)

...

offset_mapping = out.pop("offset_mapping")
input_ids = out["input_ids"]

for assistant_start_char, assistant_end_char in generation_indices[i]:
    start_pos = bisect.bisect_left(offset_starts, assistant_start_char)
    end_pos = bisect.bisect_left(offset_starts, assistant_end_char)
```
My understanding is:
- generation_indices is computed from the rendered text prompt returned by render_jinja_template
- but in multimodal processing, extra placeholder tokens such as <|image_pad|> are inserted later by the processor/tokenizer path
- therefore offset_mapping corresponds to the expanded multimodal text, while generation_indices still refers to the pre-expanded text
- this makes the character spans misaligned, so the assistant span lookup fails and assistant_masks ends up all zeros

### PR DESCRIPTION
**Pull request by Rocketknight1**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48793) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48793)

Expanding image tokens in processors was breaking the masks from `return_assistant_mask`. This PR fixes it, so hopefully they work in future! Also adds a small test (and skips a couple of smaller models where that test doesn't work)

Fixes #44521

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @zucchini-nlp here's the agent PR for #44521! I trimmed it a bit and the logic looks good to me, but I might still be overlooking something.

File: tests/test_processing_common.py [hunk 1]
```diff
@@ -1563,16 +1563,37 @@ def test_apply_chat_template_assistant_mask(self):
...
+            token_id
+            for token_id, is_assistant in zip(reference["input_ids"][0], reference["assistant_masks"][0])
+            if is_assistant
+        ]
+
+        # Regression test for #44521: expanding the placeholder into N image tokens must not shift the assistant spans
+        image_token = getattr(self, "image_token", None)
+        if image_token and self.does_processor_return_mm_offsets(processor.__class__, "replace_image_token"):
+            messages[0][0]["content"].insert(0, {"type": "image", "image": self.prepare_images_inputs()})
+
```
Comment (zucchini-nlp) [inline tests/test_processing_common.py]: we'd need to check several images per text, and also batched mixed inputs (image-text and text-only samples)
The char indices are tricky and easy to miss out edge cases

File: src/transformers/processing_utils.py [hunk 2]
```diff
@@ -2081,6 +2081,7 @@ def apply_chat_template(
                 processor_kwargs["return_offsets_mapping"] = (
                     True  # force offset mapping so we can infer token boundaries
                 )
+                processor_kwargs["return_text_replacement_offsets"] = True
```
Comment (zucchini-nlp) [inline src/transformers/processing_utils.py:2084]: nice, so core models with chat template all support the arg though we might break remote code - not very sure. Gotta check and think how to best support BC

File: src/transformers/processing_utils.py [hunk 3]
```diff
@@ -2227,13 +2228,30 @@ def apply_chat_template(
...
                         offsets = offset_mapping[i]
                         offset_starts = [start for start, end in offsets]
+                        placeholder_ends = [r["span"][1] for r in replacement_offsets[i]]
+                        chars_gained = [0] + [r["new_span"][1] - r["span"][1] for r in replacement_offsets[i]]
                         for assistant_start_char, assistant_end_char in generation_indices[i]:
+                            assistant_start_char, assistant_end_char = (
+                                char + chars_gained[bisect.bisect_right(placeholder_ends, char)]
+                                for char in (assistant_start_char, assistant_end_char)
+                            )
                             start_pos = bisect.bisect_left(offset_starts, assistant_start_char)
```
Comment (zucchini-nlp) [inline src/transformers/processing_utils.py]: ahh i see, so we do have code for multiple `mm_data` in single text, though a test wouldn't hurt. First time I added `return_text_replacement_offsets` it had some hidden bugs with several images per text, so i want us to  carefully test 😢

**Review by zucchini-nlp — COMMENTED** [review]
I think the rest is better tested by a human for edge cases but the general idea is indeed like I had in my mind

**Comment by Rocketknight1** [comment]
@zucchini-nlp modified the code to support padding properly, and to test batches with multiple images/samples. It's not that performant for now but probably fine, if it turns out to be a problem I can add different code paths for array/list/etc. attention masks

**Review by zucchini-nlp — APPROVED** [review]
Thanks, agreed that we can optimize the loop later but rn it is rarely used imo. I'm still concerned about passing `return_text_replacement_offsets` to processors and remote code

Let's check with vllm (it sometimes hosts its own processor classes) and maybe TRL, i don't know who else might try to get assistant masks. If they're fine, we can merge

**Comment by Rocketknight1** [comment]
I think it's okay, right? That kwarg is already being used and vLLM support shouldn't be a problem after https://github.com/huggingface/transformers/pull/47614. But if you're unsure I can leave this for a while!

**Comment by zucchini-nlp** [comment]
i meant they have a bunch of custom/remote code in https://github.com/vllm-project/vllm/tree/main/vllm/transformers_utils/processors , I'm just not sure how chat template gets applied in those remote processors (i.e. do they use assistant mask)

**Comment by Rocketknight1** [comment]
I took a look around, and I -think- this is safe. It looks like vLLM does use assistant mask, but it routes all chat template requests [through the tokenizer, not the processor](https://github.com/vllm-project/vllm/blob/44dd18fe0bb0f13157f97a5aa029b6604468fca6/vllm/renderers/hf.py#L833-L852), and then presumably does image token expansion somewhere else later. So anything we do with processors + attention_mask just won't affect them at all

**Comment by zucchini-nlp** [comment]
oke thanks for double-checking, i think i am just traumatized by vllm atp 😆
